lisa/tests/hotplug.py

```diff
@@ -269,4 +271,174 @@ def cpuhp_seq(cls, nr_operations, hotpluggable_cpus, max_cpus_off, random_gen):
+                     res_dir: ArtifactPath = None) -> 'HotplugRollback':
+        target.hotplug.online_all()
+        cpu = target.hotplug.list_hotpluggable_cpus()[0]
+        ftrace_coll = FtraceCollector.from_conf(target, cls.ftrace_conf)
+
+        # Get the list of each state that can fail
+        states_down = cls._get_states(target, cpu, up=False)
+        states_up = cls._get_states(target, cpu, up=True)
+
+        cls._prepare_hotplug(target, cpu, up=False)
+        with ftrace_coll:
+            # Get the expected list of states for a complete Hotplug
+            cls._mark_trace(target, ftrace_coll, start=1, expected=1, up=False)
+            cls._online(target, cpu, online=False)
+            cls._mark_trace(target, ftrace_coll, start=1, expected=1, up=True)
+            cls._online(target, cpu, online=True)
+            cls._mark_trace(target, ftrace_coll, start=0)
+
+            # Test hotunplug rollback for each possible state failure
+            for failing_state in states_down:
+                cls._test_rollback(target, ftrace_coll, cpu, failing_state,
+                    up=False)
+
+            # Test hotplug rollback for each possible state failure
+            for failing_state in states_up:
+                cls._test_rollback(target, ftrace_coll, cpu, failing_state,
+                    up=True)
+
+        ftrace_coll.get_data(ArtifactPath.join(res_dir, cls.TRACE_PATH))
+
+        return cls(res_dir, target.plat_info)
+
+    @classmethod
+    def _get_expected_states(cls, df, up):
+        df = df[(df['expected'] == 1.0) & (df['up'] == str(up))]
+
+        return df['idx'].dropna().tolist()
+
+    @requires_events('userspace@HotplugRollback', 'cpuhp_enter')
+    def test_hotplug_rollback(self) -> ResultBundle:
+        """
+        Test that the hotplug can rollback to its previous state after a
+        failure. All possible steps, up/down combinations will be tested. For
+        each combination, also verify that the hotplug is going through all the
+        steps it is supposed to.
+        """
+        df = df_merge([
+            self.trace.df_event('userspace@HotplugRollback'),
+            self.trace.df_event('cpuhp_enter')
+        ])
+
+        # Keep only the states delimited by _mark_trace()
+        df['test'].ffill(inplace=True)
+        df = df[df['test'] == 1]
+
+        # Read the expected states from full hot(un)plug
+        df['expected'].ffill(inplace=True)
+        df['up'].ffill(inplace=True)
+        expected_down = self._get_expected_states(df, up=False)
+        expected_up = self._get_expected_states(df, up=True)
+
+        def _get_expected_rollback(up, failing_state):
+            if up == 'True':
+                expected = [
+                    s for s in expected_up if s < int(failing_state)]
+                expected.extend([
+                    s for s in expected_down if s < int(failing_state)
+                ])
+            else:
+                expected = [
```

Comment:
Actually, the whole function can be heavily deduplicated:
```
import operator
from itertools import chain

def _get_expected_rollback(up, failing_state):
    comp = operator.lt if up else operator.gt
    return [
        s
        for s in chain(expected_up, expected_down)
        if comp(s, failing_state)
    ]

```

Or event actually turned into a single expression
```

from functools import partial
def _get_expected_rollback(up, failing_state):
    return list(filter(
        partial(
            operator.gt if up else operator.lt,
            failing_state,
        ),
        chain(expected_up, expected_down)
    ))

```